### PR TITLE
[5.5] Fix flaws in how search_path is set on connection

### PR DESCRIPTION
**Pull request by cbj4074**
TL;DR
Currently, Laravel conflates "schema" with "search_path" and prepares statement parameter bindings incorrectly when setting the value, which could have significant negative implications for PostgreSQL-backed applications. This commit corrects the terminology and configuration, with backward compatibility.
/TL;DR

This commit seeks to address two fundamental issues: 1) incorrect PostgreSQL terminology and application of concepts; 2) ineffective "search_path" values set on connection due to incorrect prepared statement parameter bindings (entire comma-separated string of schemas is quoted when each schema must be quoted separately).

In the existing code, "schema" is conflated with "search_path", which has significant negative implications for PostgreSQL-backed applications.

This commit corrects the terminology within the source so that the code reflects more accurately its actual behavior, but more importantly, it changes the connection configuration variable name that is used to set the "search_path" from "schema" to "searchpath".

The change allows for "decoupling" in that "schema" can still be used for its intended purpose (to prefix object references for schemas not in the "search_path", or to disambiguate objects of the same name that exist in more than one schema in the "search_path"), while "searchpath" can now be used for *its* intended purpose. In other words, "schema" and "searchpath" are now two separate and unrelated directives, which can be configured independently.

Note that the "schema" connection property is no longer used when configuring the connection, but may still be accessed as needed via getConnection()->getConfig('schema').

The change is backward-compatible in that if the "searchpath" key is not defined on the connection, the "schema" key will be used to set the search_path instead. If "schema" is not defined, PostgreSQL's default schema ("public") will be used.

Note further that "searchpath" may be specified as a string (which allows for a single schema only), or an array (which allows for any number of schemas, excluding only dynamic schemas that have special meaning to PostgreSQL, such as '$user'; support for $-style variable notation may be added in a future revision).

For an exhaustive discussion, see: https://github.com/laravel/internals/issues/918

### CODE REVIEW

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 1]
```diff
@@ -81,33 +81,41 @@ protected function configureTimezone($connection, array $config)
...
+            $schemas = array_map(function($item) {
+                return '"'.$item.'"';
+            }, $searchPath);
+
+            $preparedSchemas = array_map(function($item) {
+                return ":$item";
+            }, $schemas);
+
+            return implode(', ', array_combine($preparedSchemas, $schemas));
         } else {
```
Comment (jmarcher) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: this else statement is unnecessary

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 2]
```diff
@@ -81,33 +81,41 @@ protected function configureTimezone($connection, array $config)
...
      * @param  array  $config
      * @return void
      */
-    protected function configureSchema($connection, $config)
+    protected function configureSearchPath($connection, $config)
     {
-        if (isset($config['schema'])) {
-            $schema = $this->formatSchema($config['schema']);
+        if (isset($config['searchpath']) || isset($config['schema'])) {
+            $searchPath = $this->formatSearchPath($config['searchpath'] ?? $config['schema'] ?? '"public"');
```
Comment (jmarcher) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: This line looks a bit weird with the chain of ??

**Review by jmarcher — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 2 above)
Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: I'm sorry can you elaborate on "looks weird" ?

Comment (jmarcher) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: It looks like a new protected method would work better than chaining the null coalesce operator

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 1 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Not if we want the existing tests to pass...

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 2 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Can you elaborate on "better"? Hehe... seriously, though, this seems like the "simplest" solution to me. If there is some worthwhile advantage to be gained by adding a new method, then feel free to propose the change.

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Not to sound smug but, "work better" ? It would work no differently then what is here ? If you are referring to readability, then the change is merely an option. This implementation does not violate any coding principles or practices. Please be more clear with your comments and maybe contribute with an example.

Thanks in advance

@jmarcher

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 1 above)
Comment (dlett) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: I believe what he means is that you could keep the code inside of the else block but remove the else keyword itself and the functionality would remain the same. eg:

```
if (is_array($searchPath)) {
    ...
    return implode(', ', array_combine($preparedSchemas, $schemas));
}

return implode(', ', [":$searchPath" => '"'.$searchPath.'"']);
```

@cbj4074

Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: @dlett Aha! Of course that's what he meant. Been a long few days... thanks!

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: This is exactly the reason to provide more clear suggestions and examples. Thanks for clearing that up @dlett . I do agree that this style makes more since.

However, this is the style preference from the original ....

```php
/**
     * Format the schema for the DSN.
     *
     * @param  array|string  $schema
     * @return string
     */
    protected function formatSchema($schema)
    {
        if (is_array($schema)) {
            return '"'.implode('", "', $schema).'"';
        } else {
            return '"'.$schema.'"';
        }
    }
```

Reference: https://github.com/laravel/framework/blob/5.5/src/Illuminate/Database/Connectors/PostgresConnector.php#L109

Not sure why this isn't merged yet to be honest. 

Any serious discussion in regards to why this should not be merged ??

Comment (taylorotwell) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Things take a little time to get merged sometimes. There is no specific reason. It's just nature of open source :smile:

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: @taylorotwell sure, I understand that completely.

My concern was the hair splitting discussions around personal style preferences with no relevant discussions on actual concerns with the ability to merge this proposal. I have no issues making changes to satisfy any concerns, but they should be justified.

I don't mean to offend anyone involved. Just my 2 cents ;)

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 2 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: And this is precisely why I left it alone: https://github.com/laravel/framework/pull/22425

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 1 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: And this is precisely why I left it alone: https://github.com/laravel/framework/pull/22425

Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Hopefully, everybody is happy now. You win, @jmarcher and @carusogabriel . :clap:

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 3]
```diff
@@ -81,34 +81,42 @@ protected function configureTimezone($connection, array $config)
...
```
Comment (mfn) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: @cbj4074 since you introduced `->formatSearchPath()`, doesn't this mean the last null-coalesce default value should be simply `'public'` instead of `'"public"'`, as the method will take care of it?

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: You may be right. It will be a few minutes but I'll fire up the laptop and confirm this.

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: @mfn Tests do check out in either case. Regarding tests they explicitly check a literal string which tells me in both cases the string is formatted correctly.

tests/Database/DatabaseConnectorTest.php
```php
        $connection->shouldReceive('prepare')->once()->with('set search_path to "public"')->andReturn($connection);
```

Comment (poppabear8883) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: In fact we could remove the `?? '"public"'` altogether and tests will pass as we explicitly check that either `schema` OR `searchpath` isset ... and tests would still pass

**Comment by cbj4074** [comment]
Merged upstream... again. Are waiting on something specific here?

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 4]
```diff
@@ -81,34 +81,42 @@ protected function configureTimezone($connection, array $config)
...
+        if (is_array($searchPath)) {
+            $schemas = array_map(function ($item) {
+                return '"'.$item.'"';
+            }, $searchPath);
+
+            $preparedSchemas = array_map(function ($item) {
+                return ":$item";
+            }, $schemas);
+
+            return implode(', ', array_combine($preparedSchemas, $schemas));
```
Comment (mfn) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: I don't understand your use of the key=>value approach together with `implode`.

In my book, the keys when performing implode are ignored?

```php
 $ php -v
PHP 7.1.11-1…
…
$ php -r 'var_dump(implode(",", ["a"=>"b", "c" => "d"]));'
string(3) "b,d"
```

I understand you try to use something with prepared statements here but IMHO it ends up being not used.

**Comment by cbj4074** [comment]
Several of us have performed additional research regarding this subject and would like to revise the PR. We'll open another one when we're ready for feedback.

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 5]
```diff
@@ -81,34 +81,42 @@ protected function configureTimezone($connection, array $config)
...
-     * @param  array|string  $schema
+     * @param  array|string  $searchPath
      * @return string
      */
-    protected function formatSchema($schema)
+    protected function formatSearchPath($searchPath)
     {
-        if (is_array($schema)) {
-            return '"'.implode('", "', $schema).'"';
+        if (is_array($searchPath)) {
```
Comment (mfn) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: Small suggestion; instead of having `is_array / else` (like the old code), I _think_ we can get rid of nesting the whole method if you actually convert the string into an array:
```php
$searchPath = (array) $searchPath;
// the code always just has to handle arrays
```

See:
```php
$ php -r 'var_dump( (array) "foo");'
array(1) {
  [0]=>
  string(3) "foo"
}

$ php -r 'var_dump( (array) ["foo"]);'
array(1) {
  [0]=>
  string(3) "foo"
}

File: src/Illuminate/Database/Connectors/PostgresConnector.php [hunk 6]
```diff
@@ -81,34 +81,42 @@ protected function configureTimezone($connection, array $config)
...
      * @param  \PDO  $connection
      * @param  array  $config
      * @return void
      */
-    protected function configureSchema($connection, $config)
+    protected function configureSearchPath($connection, $config)
     {
-        if (isset($config['schema'])) {
-            $schema = $this->formatSchema($config['schema']);
+        if (isset($config['searchpath']) || isset($config['schema'])) {
```
Comment (mfn) [inline src/Illuminate/Database/Connectors/PostgresConnector.php]: Any reason why the config should be named `searchpath` when the postgres run-time parameter is called `search_path`?

E.g. the same class also deals with `application_name` which is named that way _in_ the configuration and in postgres.

**Review by mfn — COMMENTED** [review]
Guess I'm too late to the game as it was already closed in the meantime 😄 

I'm definitely pro fixing this (postgres user here, though not using non-standard `search_path` _at the moment_).

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 6 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: None at all! I had overlooked the fact that `application_name` contains an underscore, and concur that it should perhaps be `search_path`, for the sake of consistency. Nice catch!

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 5 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Precisely. That is one of the things we are seeking to change.

File: src/Illuminate/Database/Connectors/PostgresConnector.php (hunk 4 above)
Comment (cbj4074) [reply src/Illuminate/Database/Connectors/PostgresConnector.php]: Again, you are spot-on!

**Comment by cbj4074** [comment]
@mfn Where were you three days ago?! 🤣 Seriously, though, your insights are succinct, helpful, and very much appreciated. This PR simply grew to be too messy and confusing, so we're paring-down and working to present the perceived limitations and proposed changes as straightforwardly as possible. We'll keep you in the loop and would love for you to take a look as soon as it's posted.

**Comment by mfn** [comment]
Thanks, appreciated being kept in the loop!

One more thing: by accident I've came over `\Illuminate\Database\Schema\PostgresBuilder::parseSchemaAndTable`:
```php
```

**Comment by poppabear8883** [comment]
@mfn is there any tests that touch that method ? I can't seem to find any. I am assuming that logically this method would work with our *new* implementation as tests are all passing. So are you referring to updating it to be consistent with naming specifically ?

**Comment by mfn** [comment]
I might have missed something, but the pasted code uses:
```php
$this->connection->getConfig('schema')
```
And given the attempt in this PR, yes, `schema` would still be supported but "the new way" would be the `search_path` variable which isn't covered in this codepath and might yield unexpected results therefore, methinks.

**Comment by poppabear8883** [comment]
Excellent point with the `getConfig('schema')`. I totally over looked that. 

You are right to assume that, it in fact would need to be updated.

Thanks
